### ISSUE TITLE
Uncaught TypeError: Cannot read property '0' of null

### ISSUE THREAD

**Issue by handlerug** [atom/atom#17023]
_Edited by @ben3eee to include steps to reproduce from https://github.com/atom/atom/issues/17023#issuecomment-376414075_

1. Open Atom in safe mode `atom --safe`
1. Open a new untitled tab
1. Change the language mode to HTML
1. Change line endings to be `CRLF`
1. Paste the following contents:
```html
<script type="text/javascript">
  if (Math.abs(resVal - resArrayUpper[lowerBound].resistance) <
      Math.abs(resVal - resArrayUpper[upperBound].resistance)) {
    return lowerBound;
  } else {
    return upperBound;
  }
</script>
```
5. Place the cursor on line 7 and try to add a newline

**Expected:** Atom to add a newline. Editing to work as normal
**Actual:** This exception is thrown and no newline is added

If you add a whitespace at the end of line 2 editing works as normal

![property 0 of null](https://user-images.githubusercontent.com/1058982/37950800-44a64fec-319a-11e8-98f3-dd86c1fd4684.gif)

---

[Enter steps to reproduce:]

1. Paste a long line of text in HTML file
2. Start typing a \<p\> tag in beginning of this line

**Atom**: 1.25.0 x64
**Electron**: 1.7.11
**OS**: Mac OS X 10.13.3
**Thrown From**: Atom Core

### Stack Trace

Uncaught TypeError: Cannot read property '0' of null

```
At /Applications/Atom.app/Contents/Resources/app/node_modules/bracket-matcher/lib/tag-finder.js:168

TypeError: Cannot read property '0' of null
    at /Applications/Atom.app/Contents/Resources/app/node_modules/bracket-matcher/lib/tag-finder.js:168:36
    at TextBuffer.module.exports.TextBuffer.scanInRange (/Applications/Atom.app/Contents/Resources/app/node_modules/text-buffer/lib/text-buffer.js:1097:15)
    at TextBuffer.module.exports.TextBuffer.backwardsScanInRange (/Applications/Atom.app/Contents/Resources/app/node_modules/text-buffer/lib/text-buffer.js:1115:25)
    at TextEditor.backwardsScanInBufferRange (/Applications/Atom.app/Contents/Resources/app/src/text-editor.js:3392:82)
    at TagFinder.module.exports.TagFinder.findStartEndTags (/Applications/Atom.app/Contents/Resources/app/node_modules/bracket-matcher/lib/tag-finder.js:163:25)
    at TagFinder.module.exports.TagFinder.findMatchingTags (/Applications/Atom.app/Contents/Resources/app/node_modules/bracket-matcher/lib/tag-finder.js:208:27)
    at BracketMatcherView.module.exports.BracketMatcherView.updateMatch (/Applications/Atom.app/Contents/Resources/app/node_modules/bracket-matcher/lib/bracket-matcher-view.js:135:41)
    at /Applications/Atom.app/Contents/Resources/app/node_modules/bracket-matcher/lib/bracket-matcher-view.js:3:65
    at Function.module.exports.Emitter.simpleDispatch (/Applications/Atom.app/Contents/Resources/app/node_modules/event-kit/lib/emitter.js:25:20)
    at Emitter.module.exports.Emitter.emit (/Applications/Atom.app/Contents/Resources/app/node_modules/event-kit/lib/emitter.js:141:34)
    at TextBuffer.module.exports.TextBuffer.emitDidChangeTextEvent (/Applications/Atom.app/Contents/Resources/app/node_modules/text-buffer/lib/text-buffer.js:1748:32)
    at TextBuffer.module.exports.TextBuffer.transact (/Applications/Atom.app/Contents/Resources/app/node_modules/text-buffer/lib/text-buffer.js:975:18)
    at TextEditor.transact (/Applications/Atom.app/Contents/Resources/app/src/text-editor.js:1839:30)
    at mergeIntersectingSelections (/Applications/Atom.app/Contents/Resources/app/src/text-editor.js:1386:25)
    at TextEditor.mergeSelections (/Applications/Atom.app/Contents/Resources/app/src/text-editor.js:3249:53)
    at TextEditor.mergeIntersectingSelections (/Applications/Atom.app/Contents/Resources/app/src/text-editor.js:3223:42)
    at TextEditor.mutateSelectedText (/Applications/Atom.app/Contents/Resources/app/src/text-editor.js:1385:23)
    at TextEditor.insertText (/Applications/Atom.app/Contents/Resources/app/src/text-editor.js:1349:31)
    at TextEditor.object.(anonymous function) [as insertText] (/Applications/Atom.app/Contents/Resources/app/node_modules/underscore-plus/lib/underscore-plus.js:77:33)
    at TextEditorComponent.didTextInput (/Applications/Atom.app/Contents/Resources/app/src/text-editor-component.js:1654:30)
```

### Commands

```
     -2:47.9.0 autocomplete-plus:activate (atom-text-editor.editor.is-focused)
  2x -2:45.7.0 core:backspace (input.hidden-input)
     -1:17.5.0 editor:newline (input.hidden-input)
  4x -1:16.5.0 core:move-down (input.hidden-input)
     -1:15.8.0 autocomplete-plus:confirm (input.hidden-input)
     -1:15.8.0 autocomplete-plus:activate (atom-text-editor.editor.is-focused)
  4x -1:14.2.0 core:backspace (input.hidden-input)
     -1:11.7.0 core:save (input.hidden-input)
     -1:05.1.0 core:paste (input.hidden-input)
  2x -0:59 core:move-down (input.hidden-input)
     -0:58.2.0 editor:move-to-end-of-screen-line (input.hidden-input)
     -0:55.8.0 core:backspace (input.hidden-input)
     -0:55.2.0 core:move-down (input.hidden-input)
     -0:54.9.0 core:move-left (input.hidden-input)
     -0:54.3.0 editor:move-to-first-character-of-line (input.hidden-input)
  6x -0:18.5.0 core:backspace (input.hidden-input)
```

### Non-Core Packages

```
atom-ide-ui 0.9.4 
atom-material-syntax 1.0.8 
atom-material-ui 2.1.3 
city-lights-syntax 1.1.8 
city-lights-ui 1.3.3 
emmet 2.4.3 
ide-python 0.8.2 
ide-typescript 0.7.6 
language-babel 2.84.0 
language-ejs 0.4.0 
language-postcss 1.3.1 
oceanic-next 1.0.0 
teletype 0.10.0 
```

**Comment by rsese** [comment]
Thanks for the report - can you reproduce the error in safe mode (`atom --safe`)?

**Comment by slugrustle** [comment]
Same error here.

[Enter steps to reproduce:]

This crops up frequently while editing the file high_voltage_divider.html from the project [https://github.com/slugrustle/high_voltage_divider](https://github.com/slugrustle/high_voltage_divider)

I will try it in safe mode.

**Atom**: 1.25.0 x64
**Electron**: 1.7.11
**OS**: Microsoft Windows 10 Home
**Thrown From**: Atom Core


### Stack Trace

Uncaught TypeError: Cannot read property '0' of null

```
At C:\Users\user\AppData\Local\atom\app-1.25.0\resources\app\node_modules\text-buffer\lib\text-buffer.js:959

TypeError: Cannot read property '0' of null
    at ~/AppData/Local/atom/app-1.25.0/resources/app/node_modules/bracket-matcher/lib/tag-finder.js:168:36
    at TextBuffer.module.exports.TextBuffer.scanInRange (~/AppData/Local/atom/app-1.25.0/resources/app/node_modules/text-buffer/lib/text-buffer.js:1097:15)
    at TextBuffer.module.exports.TextBuffer.backwardsScanInRange (~/AppData/Local/atom/app-1.25.0/resources/app/node_modules/text-buffer/lib/text-buffer.js:1115:25)
    at TextEditor.backwardsScanInBufferRange (~/AppData/Local/atom/app-1.25.0/resources/app/src/text-editor.js:3392:82)
    at TagFinder.module.exports.TagFinder.findStartEndTags (~/AppData/Local/atom/app-1.25.0/resources/app/node_modules/bracket-matcher/lib/tag-finder.js:163:25)
    at TagFinder.module.exports.TagFinder.findMatchingTags (~/AppData/Local/atom/app-1.25.0/resources/app/node_modules/bracket-matcher/lib/tag-finder.js:208:27)
    at BracketMatcherView.module.exports.BracketMatcherView.updateMatch (~/AppData/Local/atom/app-1.25.0/resources/app/node_modules/bracket-matcher/lib/bracket-matcher-view.js:135:41)
    at ~/AppData/Local/atom/app-1.25.0/resources/app/node_modules/bracket-matcher/lib/bracket-matcher-view.js:3:65
    at Function.module.exports.Emitter.simpleDispatch (~/AppData/Local/atom/app-1.25.0/resources/app/node_modules/event-kit/lib/emitter.js:25:20)
    at Emitter.module.exports.Emitter.emit (~/AppData/Local/atom/app-1.25.0/resources/app/node_modules/event-kit/lib/emitter.js:141:34)
    at TextEditor.selectionRangeChanged (~/AppData/Local/atom/app-1.25.0/resources/app/src/text-editor.js:3323:24)
    at Selection.markerDidChange (~/AppData/Local/atom/app-1.25.0/resources/app/src/selection.js:937:23)
    at Selection.marker.onDidChange.e (~/AppData/Local/atom/app-1.25.0/resources/app/src/selection.js:21:45)
    at Function.module.exports.Emitter.simpleDispatch (~/AppData/Local/atom/app-1.25.0/resources/app/node_modules/event-kit/lib/emitter.js:25:20)
    at Emitter.module.exports.Emitter.emit (~/AppData/Local/atom/app-1.25.0/resources/app/node_modules/event-kit/lib/emitter.js:141:34)
    at DisplayMarker.module.exports.DisplayMarker.notifyObservers (~/AppData/Local/atom/app-1.25.0/resources/app/node_modules/text-buffer/lib/display-marker.js:248:33)
    at ~/AppData/Local/atom/app-1.25.0/resources/app/node_modules/text-buffer/lib/display-marker.js:53:32
    at Function.module.exports.Emitter.simpleDispatch (~/AppData/Local/atom/app-1.25.0/resources/app/node_modules/event-kit/lib/emitter.js:25:20)
    at Emitter.module.exports.Emitter.emit (~/AppData/Local/atom/app-1.25.0/resources/app/node_modules/event-kit/lib/emitter.js:141:34)
    at Marker.module.exports.Marker.emitChangeEvent (~/AppData/Local/atom/app-1.25.0/resources/app/node_modules/text-buffer/lib/marker.js:441:26)
    at Marker.module.exports.Marker.update (~/AppData/Local/atom/app-1.25.0/resources/app/node_modules/text-buffer/lib/marker.js:374:18)
    at Marker.module.exports.Marker.setHeadPosition (~/AppData/Local/atom/app-1.25.0/resources/app/node_modules/text-buffer/lib/marker.js:149:25)
    at DisplayMarker.module.exports.DisplayMarker.setHeadBufferPosition (~/AppData/Local/atom/app-1.25.0/resources/app/node_modules/text-buffer/lib/display-marker.js:146:38)
    at DisplayMarker.module.exports.DisplayMarker.setHeadScreenPosition (~/AppData/Local/atom/app-1.25.0/resources/app/node_modules/text-buffer/lib/display-marker.js:154:25)
    at changePosition (~/AppData/Local/atom/app-1.25.0/resources/app/src/cursor.js:68:25)
    at Cursor.changePosition (~/AppData/Local/atom/app-1.25.0/resources/app/src/cursor.js:713:11)
    at Cursor.setScreenPosition (~/AppData/Local/atom/app-1.25.0/resources/app/src/cursor.js:67:16)
    at moveCursors.cursor (~/AppData/Local/atom/app-1.25.0/resources/app/src/text-editor.js:2519:52)
    at Array.forEach (native)
    at transact (~/AppData/Local/atom/app-1.25.0/resources/app/src/text-editor.js:2709:31)
```

### Commands

```
  2x -1:54.1.0 core:move-down (input.hidden-input)
 11x -1:53.6.0 core:move-left (input.hidden-input)
     -1:51.5.0 core:move-right (input.hidden-input)
  3x -1:51.1.0 core:backspace (input.hidden-input)
 42x -1:49.3.0 core:move-left (input.hidden-input)
  3x -1:46.9.0 core:backspace (input.hidden-input)
     -1:45.4.0 core:move-right (input.hidden-input)
     -1:45.1.0 core:move-down (input.hidden-input)
 86x -1:44.9.0 core:move-right (input.hidden-input)
  3x -1:40.9.0 core:backspace (input.hidden-input)
     -1:39.2.0 core:move-right (input.hidden-input)
 17x -1:38.8.0 core:move-down (input.hidden-input)
 31x -1:07.0 core:select-up (input.hidden-input)
  2x -1:04.7.0 core:backspace (input.hidden-input)
     -0:58.1.0 intentions:highlight (input.hidden-input)
     -0:57.9.0 core:save (input.hidden-input)
```

### Non-Core Packages

```
busy-signal 1.4.3 
intentions 1.1.5 
linter 2.2.0 
linter-csslint 2.0.0 
linter-htmlhint 1.4.0 
linter-jshint 3.1.6 
linter-ui-default 1.7.1 
```

**Comment by slugrustle** [comment]
Same thing happened in safe mode while editing the same file. I just opened it and clicked on a line. There is a lot of html and javascript in that file.

**Atom**: 1.25.0 x64
**Electron**: 1.7.11
**OS**: Microsoft Windows 10 Home
**Thrown From**: Atom Core


### Stack Trace

Uncaught TypeError: Cannot read property '0' of null

```
At C:\Users\user\AppData\Local\atom\app-1.25.0\resources\app\node_modules\text-buffer\lib\text-buffer.js:959

TypeError: Cannot read property '0' of null
    at ~/AppData/Local/atom/app-1.25.0/resources/app/node_modules/bracket-matcher/lib/tag-finder.js:168:36
    at TextBuffer.module.exports.TextBuffer.scanInRange (~/AppData/Local/atom/app-1.25.0/resources/app/node_modules/text-buffer/lib/text-buffer.js:1097:15)
    at TextBuffer.module.exports.TextBuffer.backwardsScanInRange (~/AppData/Local/atom/app-1.25.0/resources/app/node_modules/text-buffer/lib/text-buffer.js:1115:25)
    at TextEditor.backwardsScanInBufferRange (~/AppData/Local/atom/app-1.25.0/resources/app/src/text-editor.js:3392:82)
    at TagFinder.module.exports.TagFinder.findStartEndTags (~/AppData/Local/atom/app-1.25.0/resources/app/node_modules/bracket-matcher/lib/tag-finder.js:163:25)
    at TagFinder.module.exports.TagFinder.findMatchingTags (~/AppData/Local/atom/app-1.25.0/resources/app/node_modules/bracket-matcher/lib/tag-finder.js:208:27)
    at BracketMatcherView.module.exports.BracketMatcherView.updateMatch (~/AppData/Local/atom/app-1.25.0/resources/app/node_modules/bracket-matcher/lib/bracket-matcher-view.js:135:41)
    at ~/AppData/Local/atom/app-1.25.0/resources/app/node_modules/bracket-matcher/lib/bracket-matcher-view.js:3:65
    at Function.module.exports.Emitter.simpleDispatch (~/AppData/Local/atom/app-1.25.0/resources/app/node_modules/event-kit/lib/emitter.js:25:20)
    at Emitter.module.exports.Emitter.emit (~/AppData/Local/atom/app-1.25.0/resources/app/node_modules/event-kit/lib/emitter.js:141:34)
    at TextEditor.selectionRangeChanged (~/AppData/Local/atom/app-1.25.0/resources/app/src/text-editor.js:3323:24)
    at Selection.markerDidChange (~/AppData/Local/atom/app-1.25.0/resources/app/src/selection.js:937:23)
    at Selection.marker.onDidChange.e (~/AppData/Local/atom/app-1.25.0/resources/app/src/selection.js:21:45)
    at Function.module.exports.Emitter.simpleDispatch (~/AppData/Local/atom/app-1.25.0/resources/app/node_modules/event-kit/lib/emitter.js:25:20)
    at Emitter.module.exports.Emitter.emit (~/AppData/Local/atom/app-1.25.0/resources/app/node_modules/event-kit/lib/emitter.js:141:34)
    at DisplayMarker.module.exports.DisplayMarker.notifyObservers (~/AppData/Local/atom/app-1.25.0/resources/app/node_modules/text-buffer/lib/display-marker.js:248:33)
    at ~/AppData/Local/atom/app-1.25.0/resources/app/node_modules/text-buffer/lib/display-marker.js:53:32
    at Function.module.exports.Emitter.simpleDispatch (~/AppData/Local/atom/app-1.25.0/resources/app/node_modules/event-kit/lib/emitter.js:25:20)
    at Emitter.module.exports.Emitter.emit (~/AppData/Local/atom/app-1.25.0/resources/app/node_modules/event-kit/lib/emitter.js:141:34)
    at Marker.module.exports.Marker.emitChangeEvent (~/AppData/Local/atom/app-1.25.0/resources/app/node_modules/text-buffer/lib/marker.js:441:26)
    at Marker.module.exports.Marker.update (~/AppData/Local/atom/app-1.25.0/resources/app/node_modules/text-buffer/lib/marker.js:374:18)
    at Marker.module.exports.Marker.setHeadPosition (~/AppData/Local/atom/app-1.25.0/resources/app/node_modules/text-buffer/lib/marker.js:149:25)
    at DisplayMarker.module.exports.DisplayMarker.setHeadBufferPosition (~/AppData/Local/atom/app-1.25.0/resources/app/node_modules/text-buffer/lib/display-marker.js:146:38)
    at DisplayMarker.module.exports.DisplayMarker.setHeadScreenPosition (~/AppData/Local/atom/app-1.25.0/resources/app/node_modules/text-buffer/lib/display-marker.js:154:25)
    at changePosition (~/AppData/Local/atom/app-1.25.0/resources/app/src/cursor.js:68:25)
    at Cursor.changePosition (~/AppData/Local/atom/app-1.25.0/resources/app/src/cursor.js:713:11)
    at Cursor.setScreenPosition (~/AppData/Local/atom/app-1.25.0/resources/app/src/cursor.js:67:16)
    at moveCursors.cursor (~/AppData/Local/atom/app-1.25.0/resources/app/src/text-editor.js:2519:52)
    at Array.forEach (native)
    at transact (~/AppData/Local/atom/app-1.25.0/resources/app/src/text-editor.js:2709:31)
```

### Commands

```
```

### Non-Core Packages

```

```

**Comment by Ben3eeE** [comment]
Thank you @slugrustle :bowing_man: I can reproduce the issue with this file!

Seems to be because of line 665. If you add a space at the end of that line or don't split it up into multiple lines you can edit the file as normal again.

1. Open Atom in safe mode `atom --safe`
1. Open a new untitled tab
1. Change the language mode to HTML
1. Paste the following contents:
```html
<script type="text/javascript">
  if (Math.abs(resVal - resArrayUpper[lowerBound].resistance) <
      Math.abs(resVal - resArrayUpper[upperBound].resistance)) {
    return lowerBound;
  } else {
    return upperBound;
  }
</script>
```
5. Place the cursor on line 7 and try to add a newline

**Expected:** Atom to add a newline. Editing to work as normal
**Actual:** This exception is thrown and no newline is added

If you add a whitespace at the end of line 2 editing works as normal

![property 0 of null](https://user-images.githubusercontent.com/1058982/37950800-44a64fec-319a-11e8-98f3-dd86c1fd4684.gif)

**Comment by Ben3eeE** [comment]
I edited the steps to reproduce in the body of the issue.

**Comment by rsese** [comment]
Couldn't reproduce on macOS 10.12.6 with either Atom 1.24.1 or 1.25 but on Windows 10, I could reproduce with 1.25 but not on 1.24.1.

**Comment by Ben3eeE** [comment]
@rsese Ah, that seems to be because of line endings. macOS defaults to LF and this does not reproduce without CRLF line endings. I'll update my steps to reproduce.

**Comment by slugrustle** [comment]
A million monkeys at a million keyboards won't reproduce Shakespeare, but they will find bugs in your text editor ;)

**Comment by smashwilson** [comment]
Fixed by atom/superstring#60, included in 553dec3e. 🌈
